internal/controller/controllers/agentserviceconfig_controller_test.go

```diff
@@ -0,0 +1,168 @@
+	testAgentServiceConfigKind       = "testKind"
+	testAgentServiceConfigAPIVersion = "testAPIVersion"
+)
+
+func newTestReconciler(initObjs ...runtime.Object) *AgentServiceConfigReconciler {
+	c := fakeclient.NewFakeClientWithScheme(scheme.Scheme, initObjs...)
+	return &AgentServiceConfigReconciler{
+		Client:    c,
+		Scheme:    scheme.Scheme,
+		Log:       ctrl.Log.WithName("testLog"),
+		Namespace: testNamespace,
+	}
+}
+
+var _ = Describe("agentserviceconfig_controller reconcile", func() {
+	var (
+		asc             *aiv1beta1.AgentServiceConfig
+		ascr            *AgentServiceConfigReconciler
+		ctx             = context.Background()
+		mockCtrl        *gomock.Controller
+		privateKey      = "test-private-key"
+		publicKey       = "test-public-key"
+		localAuthSecret *corev1.Secret
+		route           *routev1.Route
+	)
+
+	BeforeEach(func() {
+		mockCtrl = gomock.NewController(GinkgoT())
+		localAuthSecret = &corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      agentLocalAuthSecretName,
+				Namespace: testNamespace,
+			},
+			StringData: map[string]string{
+				"ec-private-key.pem": privateKey,
+				"ec-public-key.pem":  publicKey,
+			},
+			Type: corev1.SecretTypeOpaque,
+		}
+
+		asc = newDefaultAgentServiceConfig()
+		ascr = newTestReconciler(asc)
+		Expect(ascr.Client.Create(ctx, localAuthSecret)).To(BeNil())
+
+		// AgentServiceConfig created route is missing Host
+		route, _ = ascr.newAgentRoute(asc)
+		route.Spec.Host = "testHost"
+		Expect(ascr.Client.Create(ctx, route)).To(BeNil())
+	})
+
+	AfterEach(func() {
+		mockCtrl.Finish()
+		Expect(ascr.Client.Delete(ctx, asc)).ShouldNot(HaveOccurred())
+		Expect(ascr.Client.Delete(ctx, localAuthSecret)).ShouldNot(HaveOccurred())
+		Expect(ascr.Client.Delete(ctx, route)).ShouldNot(HaveOccurred())
+	})
+
+	Describe("reconcile local auth secret", func() {
+
+		Context("with an existing local auth secret", func() {
+			It("should not modify existing keys", func() {
+				result, err := ascr.Reconcile(newAgentServiceConfigRequest(asc))
+				Expect(err).To(BeNil())
+				Expect(result).To(Equal(ctrl.Result{}))
+
+				found := &corev1.Secret{}
+				err = ascr.Client.Get(context.TODO(), types.NamespacedName{Name: agentLocalAuthSecretName, Namespace: testNamespace}, found)
+				Expect(err).To(BeNil())
+
+				Expect(found.StringData["ec-private-key.pem"]).To(Equal(privateKey))
+				Expect(found.StringData["ec-public-key.pem"]).To(Equal(publicKey))
+			})
+		})
+
+		Context("with no existing local auth secret", func() {
+			It("should create new keys and not overwrite them in subsequent reconciles", func() {
+				Expect(ascr.Client.Delete(ctx, localAuthSecret)).ShouldNot(HaveOccurred())
+
+				result, err := ascr.Reconcile(newAgentServiceConfigRequest(asc))
+				Expect(err).To(BeNil())
+				Expect(result).To(Equal(ctrl.Result{}))
+
+				found := &corev1.Secret{}
+				err = ascr.Client.Get(context.TODO(), types.NamespacedName{Name: agentLocalAuthSecretName,
```

Comment:
How does this fake client work? I'm assuming it doesn't expect to have access to an actual kube API, right?
Is is using disk storage to front the objects you create or something?
Is there a better way to clean up rather than deleting each of the objects we create? This feels like a good recipe for test contamination.